cardano-cli/src/Cardano/CLI/Run/Friendly.hs

```diff
@@ -5,175 +5,190 @@
-
-friendlyAddress :: Addr StandardCrypto -> Object
-friendlyAddress addr =
-  HashMap.fromList $
-    case addr of
-      Addr net cred ref ->
-        [ "address" .=
-            object
-              [ "network" .= net
-              , "credential" .= cred
-              , "stake reference" .= textShow ref
-              , "Bech32" .= addressBech32
-              ]
-        ]
-      AddrBootstrap _ ->
-        ["bootstrap address" .= object ["Bech32" .= String addressBech32]]
-  where
-    addressBech32 =
-      case fromShelleyAddr @Api.ShelleyEra addr of
-        AddressInEra (ShelleyAddressInEra _) a -> serialiseAddress a
-        AddressInEra ByronAddressInAnyEra a -> serialiseAddress a
-
-assertObject :: HasCallStack => Value -> Object
-assertObject = \case
-  Object obj -> obj
-  val -> panic $ "expected JSON Object, but got " <> typ
-    where
-      typ =
-        case val of
-          Array{}  -> "an Array"
-          Bool{}   -> "a Boolean"
-          Null     -> "Null"
-          Number{} -> "a Number"
-          Object{} -> "an Object"
-          String{} -> "a String"
+friendlyStakeReference :: Shelley.StakeReference crypto -> Aeson.Value
+friendlyStakeReference = \case
+  Shelley.StakeRefBase cred -> toJSON cred
+  Shelley.StakeRefNull -> Null
+  Shelley.StakeRefPtr ptr -> toJSON ptr
+
+friendlyUpdateProposal :: TxUpdateProposal era -> Aeson.Value
+friendlyUpdateProposal = \case
+  TxUpdateProposalNone -> Null
+  TxUpdateProposal _ p -> String $ textShow p
+
+friendlyCertificates :: TxCertificates ViewTx era -> Aeson.Value
+friendlyCertificates = \case
+  TxCertificatesNone    -> Null
+  TxCertificates _ cs _ -> toJSON $ map textShow cs
+
+friendlyFee :: TxFee era -> Aeson.Value
+friendlyFee = \case
+  TxFeeImplicit _     -> "implicit"
+  TxFeeExplicit _ fee -> friendlyLovelace fee
+
+friendlyLovelace :: Lovelace -> Aeson.Value
+friendlyLovelace (Lovelace value) = String $ textShow value <> " Lovelace"
+
+friendlyMintValue :: TxMintValue ViewTx era -> Aeson.Value
+friendlyMintValue = \case
+  TxMintNone        -> Null
+  TxMintValue _ v _ -> toJSON v
+
+friendlyTxOutValue :: TxOutValue era -> Aeson.Value
+friendlyTxOutValue = \case
+  TxOutAdaOnly _ lovelace -> friendlyLovelace lovelace
+  TxOutValue _ multiasset -> toJSON multiasset
+
+friendlyMetadata :: TxMetadataInEra era -> Aeson.Value
+friendlyMetadata = \case
+  TxMetadataNone                   -> Null
+  TxMetadataInEra _ (TxMetadata m) -> toJSON $ friendlyMetadataValue <$> m
+
+friendlyMetadataValue :: TxMetadataValue -> Aeson.Value
+friendlyMetadataValue = \case
+  TxMetaNumber int   -> toJSON int
+  TxMetaBytes  bytes -> String $ textShow bytes
+  TxMetaList   lst   -> array $ map friendlyMetadataValue lst
+  TxMetaMap    m     ->
+    array
+      [array [friendlyMetadataValue k, friendlyMetadataValue v] | (k, v) <- m]
+  TxMetaText   text  -> toJSON text
+
+friendlyAuxScripts :: TxAuxScripts era -> Aeson.Value
```

Comment:
why not `friendlyAuxScripts :: CardanoEra era TxAuxScripts era -> Aeson.Value` and move `auxScriptsSupportedInEra` into this function? It will simplify `friendlyTxBody`.

Reply:
For declarativity, I wanted to have all keys in one place. `friendlyValidityRange` is the only exception because the keys depend on the era.

The style suggested by you may spread this logic into many small functions. I'm not sure this is good.

Reply:
I'm not sure I understand what you mean by "have all keys in one place". You can leave the JSON keys in place. There are already several small functions, what I suggested will remove the list comprehension noise and the additional `++` operators.

Reply:
This is what I mean:
```
diff --git a/cardano-cli/src/Cardano/CLI/Run/Friendly.hs b/cardano-cli/src/Cardano/CLI/Run/Friendly.hs
index 347c89a4e..874afa3c5 100644
--- a/cardano-cli/src/Cardano/CLI/Run/Friendly.hs
+++ b/cardano-cli/src/Cardano/CLI/Run/Friendly.hs
@@ -48,9 +48,7 @@ friendlyTxBody
         , "fee"     .= friendlyFee txFee
         , "inputs"  .= friendlyInputs txIns
         , "outputs" .= map friendlyTxOut txOuts
-        ]
-    ++  [ "auxiliary scripts" .= friendlyAuxScripts txAuxScripts
-        | Just _ <- [auxScriptsSupportedInEra era]
+        , "auxiliary scripts" .= friendlyAuxScripts era txAuxScripts
         ]
     ++  [ "certificates" .= friendlyCertificates txCertificates
         | Just _ <- [certificatesSupportedInEra era]
@@ -185,8 +183,17 @@ friendlyMetadataValue = \case
       [array [friendlyMetadataValue k, friendlyMetadataValue v] | (k, v) <- m]
   TxMetaText   text  -> toJSON text
 
-friendlyAuxScripts :: TxAuxScripts era -> Aeson.Value
-friendlyAuxScripts = \case
+friendlyAuxScripts :: CardanoEra era -> TxAuxScripts era -> Aeson.Value
+friendlyAuxScripts era txAuxScripts =
+  case auxScriptsSupportedInEra era of
+    Nothing -> Aeson.Null
+    Just _ ->
+      case txAuxScripts of
+        TxAuxScriptsNone -> Null
+        TxAuxScripts _ scripts -> String $ textShow scripts
+
+friendlyAuxScripts' :: TxAuxScripts era -> Aeson.Value
+friendlyAuxScripts' = \case
   TxAuxScriptsNone       -> Null
   TxAuxScripts _ scripts -> String $ textShow scripts
 
```

Reply:
My proposition gives more information, `null` has a meaning of "this field might be filled but is not", and an omitted field means "this cannot be filled".

> we should render `Null` in the case when it's not supported

I don't understand why should we render any value for a field when it has no sense. Why should we drop this information?

Reply:
We can put all values in a single list, wrapping in `Unsupported | Supported a` datatype, and this filter out unsupported.

Reply:
We should be explicit with the unsupported fields. We are explicit in other JSON renderings in the cli and report `Null` for those cases where we do not have information to render.

Reply:
done